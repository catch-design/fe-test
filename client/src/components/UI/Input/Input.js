import React from 'react';
import classes from './Input.scss';

const input = ( props ) => {
    let inputElement = null;
    const inputClasses = [classes.signupform__inputelement];

    if(props.elementType === 'select'){
        inputClasses.push(classes.signup__select);
    }else{
        inputClasses.push(classes.signup__inputtext);
    }

    if (props.invalid && props.shouldValidate && props.touched) {
        inputClasses.push(classes.invalid);
    }

    switch ( props.elementType ) {
        case ( 'input' ):
            inputElement = <input
                className={inputClasses.join(' ')}
                {...props.elementConfig}
                value={props.value}
                onChange={props.changed} />;
            break;
        case ( 'select' ):
            inputElement = (
                <select
                    className={inputClasses.join(' ')}
                    value={props.value}
                    onChange={props.changed}>
                    {props.elementConfig.options.map(option => (
                        <option key={option.value} value={option.value}>
                            {option.name}
                        </option>
                    ))}
                </select>
            );
            break;
        default:
            inputElement = <input
                className={inputClasses.join(' ')}
                {...props.elementConfig}
                value={props.value}
                onChange={props.changed} />;
    }

    return (
        <div className={classes.signupform__inputholder}>
            {inputElement}
        </div>
    );

};

export default input;