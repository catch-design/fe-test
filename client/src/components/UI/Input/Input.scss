@import '../../../main';

@media screen and (min-width: 640px) {
  .signupform__inputholder {
    width: 50%;
    display: inline-block;
  }
}

@media screen and (max-width: 639px) {
  width: 100%;
  display: block;
}

.signupform__inputelement {
  border: 1px solid #D2DCE8;
  background-color: #ffffff;
  margin: 12px 0;
  padding: 12px;
  box-sizing: border-box;
  border-radius: 5px;;
  width: 100%;
  letter-spacing: 0.23px;
  color: $default-text-color;
  font-size: 14px;

  &.invalid {
    border: 1px solid red;
  }
}

.signupform__inputelement:focus {
  outline: none;
  background-color: #ffffff;
  border: 1px solid $default-blue;
  border-radius: 5px;
  color: $default-blue;
}

.signupform__inputelement::placeholder { /* Chrome, Firefox, Opera, Safari 10.1+ */
  letter-spacing: 0.23px;
  color: $default-text-color;
  font-size: 14px;
  opacity: 1; /* Firefox */
}

.signupform__inputelement:-ms-input-placeholder { /* Internet Explorer 10-11 */
  letter-spacing: 0.23px;
  color: $default-text-color;
  font-size: 14px;
}

.signupform__inputelement::-ms-input-placeholder { /* Microsoft Edge */
  letter-spacing: 0.23px;
  color: $default-text-color;
  font-size: 14px;
}

.signupform__inputelement:focus::placeholder { /* Chrome, Firefox, Opera, Safari 10.1+ */
  font-size: 11px;
  color: $default-text-color;
  letter-spacing: 0.18px;
  text-align: right;
  opacity: 1; /* Firefox */
}

.signupform__inputelement:focus:-ms-input-placeholder { /* Chrome, Firefox, Opera, Safari 10.1+ */
  opacity: 1; /* Firefox */
  font-size: 11px;
  color: $default-text-color;
  letter-spacing: 0.18px;
  text-align: right;
}

.signupform__inputelement:focus::-ms-input-placeholder { /* Chrome, Firefox, Opera, Safari 10.1+ */
  opacity: 1; /* Firefox */
  font-size: 11px;
  color: $default-text-color;
  letter-spacing: 0.18px;
  text-align: right;
}


.signup__select{
  height:43px;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  -o-appearance: none;
  appearance: none;

  background: url('../../../assets/images/icons/dropdown-arrow.png') no-repeat 96% 50%;
}






