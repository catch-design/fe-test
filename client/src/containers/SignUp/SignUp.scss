@import '../../main';

@media screen and (min-width: 640px) {
  .signup__container {
    width: $grid-columns-8;
    margin: 0px auto;
  }

  .signup__form {
    > :nth-child(odd) {
      padding-right: 15px;
    }

    > :nth-child(even) {
      padding-left: 15px;
    }
  }
}

@media screen and (max-width: 639px) {
  .signup__form {
    display: -ms-flexbox;
    display: flex;
    -ms-flex-direction: column;
    flex-direction: column;
  }

  .signup__container {
    padding: 0px 10px;
  }
}

.signup__container {
  text-align: center;
  max-width: 780px;
}

