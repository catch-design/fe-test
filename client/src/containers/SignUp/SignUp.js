import React, {Component} from 'react';
import { cities } from '../../models/cities';
import Input from '../../components/UI/Input/Input';
import Button from '../../components/UI/Button/Button';
import Wrapper from '../../hoc/Wrapper/Wrapper';
import classes from './SignUp.scss';
import mainStyle from '../../main.scss';

class SignUp extends Component {

    state = {
        signupForm: {
            firstName: {
                elementType: 'input',
                elementConfig: {
                    type: 'text',
                    placeholder: 'First name'
                },
                value: '',
                validation: {
                    required: true
                },
                valid: false,
                touched: false
            },
            lastName: {
                elementType: 'input',
                elementConfig: {
                    type: 'text',
                    placeholder: 'Last name'
                },
                value: '',
                validation: {
                    required: true
                },
                valid: false,
                touched: false
            },
            email: {
                elementType: 'input',
                elementConfig: {
                    type: 'email',
                    placeholder: 'Email address',
                },
                value: '',
                validation: {
                    required: true,
                    isEmail: true
                },
                valid: false,
                touched: false
            },
            city: {
                elementType: 'select',
                elementConfig: {
                    type: 'select',
                    options: cities,
                    placeholder: 'City'
                },
                validation: {
                    required: true
                },
                valid: true,
                value: cities[0].name
            }
        },
        formIsValid: false
    };

    checkValidity(value, rules) {
        let isValid = true;
        if (!rules) {
            return true;
        }

        if (rules.required) {
            isValid = value.trim() !== '' && isValid;
        }

        if (rules.isEmail) {
            const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
            isValid = pattern.test(value) && isValid
        }

        return isValid;
    }

    inputChangedHandler = (event, inputIdentifier) => {

        const updatedSignupForm = {
            ...this.state.signupForm
        };
        const updatedFormElement = {
            ...updatedSignupForm[inputIdentifier]
        };

        updatedFormElement.value = event.target.value;


        updatedFormElement.valid = this.checkValidity(updatedFormElement.value, updatedFormElement.validation);
        updatedFormElement.touched = true;
        updatedSignupForm[inputIdentifier] = updatedFormElement;

        let formIsValid = true;
        for (let inputIdentifier in updatedSignupForm) {
            formIsValid = updatedSignupForm[inputIdentifier].valid && formIsValid;
        }

        this.setState({signupForm: updatedSignupForm, formIsValid: formIsValid});
    };


    submitHandler = ( event ) => {
        event.preventDefault();
        const formData = {};
        for (let formElementIdentifier in this.state.signupForm) {
            formData[formElementIdentifier] = this.state.signupForm[formElementIdentifier].value;
        }
        const submitObject = {
            firstName: this.props.firstName,
            lastName: this.props.lastName,
            email: this.props.email,
            city: this.props.city
        }
    };

    render () {
        const formElementsArray = [];

        for (let key in this.state.signupForm) {
            formElementsArray.push({
                id: key,
                config: this.state.signupForm[key]
            });
        }

        let form = [
            <form key="form" className={classes.signup__form} onSubmit={this.submitHandler}>
                {formElementsArray.map(formElement => (
                    <Input
                        key={formElement.id}
                        elementType={formElement.config.elementType}
                        elementConfig={formElement.config.elementConfig}
                        value={formElement.config.value}
                        invalid={!formElement.config.valid}
                        shouldValidate={formElement.config.validation}
                        touched={formElement.config.touched}
                        changed={(event) => this.inputChangedHandler(event, formElement.id)} />
                ))}
            </form>,
            <Button key="submitButton" disabled={!this.state.formIsValid}>Submit</Button>
        ];
        if ( this.props.loading ) {
            form = <Spinner />;
        }
        return (
            <Wrapper>
                <div className={[classes.signup__container, mainStyle['col-8-12']].join(' ')}>
                {form}
                </div>
            </Wrapper>
        );
    }
};

export default SignUp;
