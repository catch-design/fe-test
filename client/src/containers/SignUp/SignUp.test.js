import React from 'react';
import ReactDOM from 'react-dom';
import SignUp from './SignUp';
import Input from '../../components/UI/Input/Input';
import {configure, mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

configure({adapter: new Adapter()});

const updateInput = (wrapper, instance, newValue) => {
    const input = wrapper.find(instance);
    input.simulate('change', {
        target: {value: newValue}
    });
    return wrapper.find(instance);
};

describe('<Form />', () => {
    test('allows user to fill out form', () => {
        const wrapper = mount(<SignUp/>);
        const firstNameInput = updateInput(wrapper, '[data-testid="firstName"]', 'User');
        const lastNameInput = updateInput(wrapper, '[data-testid="lastName"]', 'van User');
        const emaiIInput = updateInput(wrapper, '[data-testid="email"]', 'test@gmail.com');
        const cityInput = updateInput(wrapper, '[data-testid="email"]', 'Auckland');
        expect(firstNameInput.props().value).toBe('User');
        expect(lastNameInput.props().value).toBe('van User');
        expect(emaiIInput.props().value).toBe('test@gmail.com');
        expect(cityInput.props().value).toBe('Auckland');
        wrapper.unmount();
    });

    test('email validation', () => {
        const wrapper = mount(<SignUp/>);
        updateInput(wrapper, '[data-testid="email"]', 'test');
        expect(wrapper.state().signupForm.email.valid).toBe(false);
        updateInput(wrapper, '[data-testid="email"]', 'test@test.com');
        expect(wrapper.state().signupForm.email.valid).toBe(true);
        wrapper.unmount();
    });

    test('user submits the form', () => {
        const wrapper = mount(<SignUp/>);
        wrapper.instance().callApi = jest.fn();
        updateInput(wrapper, '[data-testid="firstName"]', 'User');
        updateInput(wrapper, '[data-testid="lastName"]', 'van User');
        updateInput(wrapper, '[data-testid="email"]', 'test@gmail.com');
        updateInput(wrapper, '[data-testid="city"]', 'Auckland');
        wrapper.find('[data-testid="signupFormSubmitButton"]').simulate('click', {
            preventDefault: () => {
            }
        });
        expect(wrapper.instance().callApi).toBeCalledWith({
            firstName: 'User',
            lastName: 'van User',
            email: 'test@gmail.com',
            city: 'Auckland'
        });
        wrapper.unmount();
    });
    // layout test
    test('matches saved snapshot', () => {
        const wrapper = shallow(<SignUp />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});