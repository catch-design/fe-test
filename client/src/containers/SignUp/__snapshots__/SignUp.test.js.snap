// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> matches saved snapshot 1`] = `
<wrapper>
  <div
    className=" "
  >
    <form
      key="form"
    >
      <input
        changed={[Function]}
        elementConfig={
          Object {
            "placeholder": "First name",
            "type": "text",
          }
        }
        elementId="firstName"
        elementType="input"
        invalid={true}
        key="firstName"
        shouldValidate={
          Object {
            "required": true,
          }
        }
        touched={false}
        value=""
      />
      <input
        changed={[Function]}
        elementConfig={
          Object {
            "placeholder": "Last name",
            "type": "text",
          }
        }
        elementId="lastName"
        elementType="input"
        invalid={true}
        key="lastName"
        shouldValidate={
          Object {
            "required": true,
          }
        }
        touched={false}
        value=""
      />
      <input
        changed={[Function]}
        elementConfig={
          Object {
            "placeholder": "Email address",
            "type": "email",
          }
        }
        elementId="email"
        elementType="input"
        invalid={true}
        key="email"
        shouldValidate={
          Object {
            "isEmail": true,
            "required": true,
          }
        }
        touched={false}
        value=""
      />
      <input
        changed={[Function]}
        elementConfig={
          Object {
            "options": Array [
              Object {
                "name": "Auckland",
                "value": "1",
              },
              Object {
                "name": "Christchurch",
                "value": "2",
              },
              Object {
                "name": "Dunedin",
                "value": "3",
              },
              Object {
                "name": "Gisborne",
                "value": "4",
              },
              Object {
                "name": "Hamilton",
                "value": "5",
              },
              Object {
                "name": "Invercargill",
                "value": "6",
              },
              Object {
                "name": "Napier-Hastings",
                "value": "7",
              },
              Object {
                "name": "Nelson",
                "value": "8",
              },
              Object {
                "name": "New Plymouth",
                "value": "9",
              },
              Object {
                "name": "Palmerston North	",
                "value": "10",
              },
              Object {
                "name": "Rotorua",
                "value": "11",
              },
              Object {
                "name": "Tauranga",
                "value": "12",
              },
              Object {
                "name": "Wellington",
                "value": "13",
              },
              Object {
                "name": "Whanganui",
                "value": "14",
              },
              Object {
                "name": "Whangarei",
                "value": "15",
              },
            ],
            "placeholder": "City",
            "type": "select",
          }
        }
        elementId="city"
        elementType="select"
        invalid={false}
        key="city"
        shouldValidate={
          Object {
            "required": true,
          }
        }
        value="Auckland"
      />
    </form>
    <button
      clicked={[Function]}
      disabled={true}
      elementId="signupFormSubmitButton"
      key="submitButton"
    >
      Submit
    </button>
  </div>
</wrapper>
`;
